### PR TITLE
Move dev15.5.x to latest IOperation analyzers for dogfooding

### PR DESCRIPTION
**Pull request by mavasani**
The primary purpose of this PR is to enable dogfooding of the latest compiler toolset, IOperation APIs and analyzers based on these APIs. The PR is for dev15.5.x rather than master because once this change is merged, Roslyn.sln developers will need one of the latest VS builds with Roslyn bits no older than from the last week for analyzers to execute fine in the IDE live analysis. This is only possible for internal developers, and master still needs to support previous VS builds for community work.

### CODE REVIEW

**Review by jasonmalinowski — APPROVED** [review]

**Comment by jasonmalinowski** [comment]
How is this being reconciled with #22380 which implied we _can't_ do this?

**Comment by jasonmalinowski** [comment]
One observation though: since this is in the dev15.5.x branch...@jaredpar what branch is being used for the source build stuff? Is it master or something else? Because having two branches might alleviate the conflict.

**Comment by mavasani** [comment]
If #22380 is a strict requirement, I think one option might be to target this change to 15.6 - I feel it should be okay for us to dogfood the new APIs on a slightly less active branch, but we do want some level of dogfooding in Jenkins. Also tagging @jinujoseph

**Comment by mavasani** [comment]
One other alternative is to make no changes or restrictions to our toolset compiler in the Roslyn repo, but instead have a separate Jenkins job that always uses a shipped toolset compiler to build the latest sources. @jaredpar will this work?

**Comment by mavasani** [comment]
I sync'ed with @jinujoseph and for now we have decided to rebase the PR to post 15.5 so we can get some dogfooding going.

**Review by jinujoseph — APPROVED** [review]

**Comment by mavasani** [comment]
@jasonmalinowski @dotnet/roslyn-infrastructure This PR is targeting post-dev15.5-contrib now. However, I see a `WIP — Waiting for status to be reported` job which doesn't seem to be triggered at all. How do I get past it or force execute it?

**Comment by jcouv** [comment]
>         <LangVersion>7.1</LangVersion>

Would you mind upgrading the language to 7.2, while you're here? Thanks

---
Refers to: build/Targets/Settings.props:238 in 75687b5. [](commit_id = 75687b52623f5873991ea5a161cbafb22c81fa82, deletion_comment = False)

**Comment by mavasani** [comment]
> Would you mind upgrading the language to 7.2, while you're here? Thanks

Done.

File: src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs
Comment (jcouv) [inline src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs]: I see this flip-flopping across multiple PRs. Does someone know where this comes from and what is the correct value moving forward?

File: build/Targets/Settings.props
Comment (jcouv) [inline build/Targets/Settings.props]: I looks like the logic to handle features (in `IsFeatureEnabled`) is being removed by this PR. Either it should stay or this last feature value can be removed.

**Comment by jcouv** [comment]
>         private ImmutableDictionary<string, string> _features;

Are there any uses of `_features` left?

---
Refers to: src/Compilers/CSharp/Portable/CSharpParseOptions.cs:22 in f5fc55f. [](commit_id = f5fc55f808fdfcfd2ff04d2fccf3b9506fb54a98, deletion_comment = False)

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Same question on `Features` here.

**Review by jcouv — COMMENTED** [review]
Done with review pass on compiler changes (iteration 4).

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb
Comment (jcouv) [reply src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Tagging @gafter who mentioned we should not remove too much features machinery, even if no features at the moment.

Comment (mavasani) [reply src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: This was actually removed a while back in dev15.5.x as it was unused code once we removed IOperation feature flag. I can revert the changes that removed the machinery with this PR.

Comment (mavasani) [reply src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Note this PR is just cherry picking existing commits that have gone in dev15.5 branch: https://github.com/dotnet/roslyn/pull/22684/commits/27dc60f5f82fff906914b7770b57ca4deaacce5c and https://github.com/dotnet/roslyn/pull/22684/commits/75687b52623f5873991ea5a161cbafb22c81fa82

File: src/Compilers/CSharp/Portable/CSharpParseOptions.cs
Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/CSharp/Portable/CSharpResources.resx
Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs
Comment (gafter) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/CSharp/Portable/Errors/MessageID.cs
Comment (gafter) [inline src/Compilers/CSharp/Portable/Errors/MessageID.cs]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/Lexer.cs]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/Test/Utilities/CSharp/TestOptions.cs
Comment (gafter) [inline src/Compilers/Test/Utilities/CSharp/TestOptions.cs]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/Test/Utilities/VisualBasic/TestOptions.vb
Comment (gafter) [inline src/Compilers/Test/Utilities/VisualBasic/TestOptions.vb]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/VisualBasic/Portable/Parser/ParserFeature.vb
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Parser/ParserFeature.vb]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb
Comment (gafter) [reply src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

**Comment by mavasani** [comment]
@gafter @jcouv I have reverted the commits that removed experimental framework. Please review again.

**Comment by mavasani** [comment]
@jcouv Ubuntu builds are failing with the below error:

```
error CS1617: Invalid option '7.2' for /langversion; must be ISO-1, ISO-2, Default, Latest or a valid version in range 1 to 7.1. [/mnt/j/workspace/dotnet_roslyn/post-dev15.5-contrib/ubuntu_14_debug_prtest/src/Compilers/Core/Portable/CodeAnalysis.csproj]
```

**Comment by jcouv** [comment]
@mavasani Ok. Take the 7.2 back. I'll investigate when I'm back in office. I suspect the problem is that we use a different package reference for .Net Core.

**Comment by jcouv** [comment]
@mavasani Anything holding up merging this?

**Comment by jcouv** [comment]
Ping @mavasani Once this is merged, I'll enable C# 7.2.

**Pull request merged**
